### PR TITLE
fix balancePercent calculations for trades on opposite side of current position

### PR DESCRIPTION
**Pull request by moo-onthelawn**
fixed a couple bugs with balancePercent calculations. math will be the end of me I swear
- Bug 1: Fix balancePercent size calculation for cross market trades on the opposite side of current position
  - The balance percent a user selects should include closing their whole current position and then apply the percentage off of their new free collateral given the entire closed position
- Bug 2: Fix balancePercent size calculation for isolated market trades on opposite side of current position
  - It doesn't make sense to apply balancePercent on the freeCollateral since it doesn't impact their free collateral - make balance percent apply to the currentPosition size instead
- Bug 3: Fix balancePercent seemingly being off for isolated market trades on same side of current position
  - It's cuz isolated markets need to do an additional calculation for the amount of additional collateral to trade when adding to current position size
  - Note: in ideal world I'd be using shared functions in MarginCalculator but I need to do a bit of refactoring to get the correct subaccount/childsubaccount variables. Filed a TODO to fix it.
  - Was able to adjust the collateral buffer percents after that fix

proof of concept help:

https://github.com/user-attachments/assets/7ec0ef78-05f4-4423-ac22-bd036f149612

### CODE REVIEW

File: src/commonMain/kotlin/exchange.dydx.abacus/calculator/V2/TradeInput/TradeInputMarketOrderCalculator.kt [hunk 1]
```diff
@@ -569,6 +633,24 @@ internal class TradeInputMarketOrderCalculator() {
...
+    private fun calculateBalancePercentFromUsdcSize(
+        usdcSize: Double,
+        freeCollateral: Double,
+        positionSize: Double,
+        tradeLeverage: Double,
+        isTradeSameSide: Boolean,
+    ): Double {
+        if (freeCollateral <= Numeric.double.ZERO) {
+            return Numeric.double.ZERO
+        }
```
Comment (jaredvu) [inline src/commonMain/kotlin/exchange.dydx.abacus/calculator/V2/TradeInput/TradeInputMarketOrderCalculator.kt:645]: would it make sense to add a non-zero check for `tradeLeverage` as well?

File: src/commonMain/kotlin/exchange.dydx.abacus/calculator/V2/TradeInput/TradeInputMarketOrderCalculator.kt [hunk 2]
```diff
@@ -558,7 +622,7 @@ internal class TradeInputMarketOrderCalculator() {
                 break@orderbookLoop
             }
         }
-        val balancePercentTotal = (usdcSizeTotal / tradeLeverage) / freeCollateral
+        val balancePercentTotal = calculateBalancePercentFromUsdcSize(usdcSize = usdcSizeTotal, freeCollateral = freeCollateral, positionSize = positionSizeNotional, tradeLeverage = tradeLeverage, isTradeSameSide = isTradeSameSide)
```
Comment (jaredvu) [inline src/commonMain/kotlin/exchange.dydx.abacus/calculator/V2/TradeInput/TradeInputMarketOrderCalculator.kt:625]: im surprised `spotless` doesn't format this to multiple lines

File: src/commonMain/kotlin/exchange.dydx.abacus/calculator/V2/TradeInput/TradeInputMarketOrderCalculator.kt [hunk 3]
```diff
@@ -118,11 +119,28 @@ internal class TradeInputMarketOrderCalculator() {
                 maxMarketLeverage
             }
 
+            val positions = subaccount.openPositions
+            val marketId = market?.perpetualMarket?.id
```
Comment (jaredvu) [inline src/commonMain/kotlin/exchange.dydx.abacus/calculator/V2/TradeInput/TradeInputMarketOrderCalculator.kt:123]: it looks like you are getting `marketId` from `trade.marketId` in the non-typed `TradeInputCalculator`. I'm assuming its interchangeable

Comment (moo-onthelawn) [reply src/commonMain/kotlin/exchange.dydx.abacus/calculator/V2/TradeInput/TradeInputMarketOrderCalculator.kt:123]: ah yeah - they should be interchangeable, but lemme make them consistent for easier to read reasons

**Review by jaredvu — APPROVED** [review]

**Pull request merged**